### PR TITLE
Move JS dependencies to correct workspaces

### PR DESCRIPTION
**Pull request by gzdunek**
This is the last step in cleaning the repository before switching to pnpm.

I moved the dependencies to workspaces where they used, or if they are need in more than one package, to the root package.json.

There are no changes to the `yarn.lock` so I didn't miss any dependencies while moving them :)

I will manually backport it to branch/v16 and (hopefully!) to v15.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -41,8 +41,58 @@
...
+    "@types/jest": "^29.5.12",
+    "@types/node": "^20.14.9",
+    "@types/react": "^18.3.3",
+    "@types/react-dom": "^18.3.0",
+    "@types/react-highlight": "^0.12.8",
+    "@types/react-router-dom": "^5.1.1",
+    "@types/react-transition-group": "^4.4.10",
+    "@types/styled-components": "^5.1.34",
+    "@types/wicg-file-system-access": "^2023.10.5",
+    "jest": "^29.7.0",
```
Comment (gzdunek) [inline package.json:57]: Initially, I wanted to keep `jest` in `packages/build`, but I didn't think about two things:
1. I need `@types/jest` globally (so TS can pick it); this suggests that `jest` should be kept in the same `package.json`.
2. Adding a script to `packages/build` and then calling it from the root,  runs only tests from `packages/build`. I'd have to change the current working directory probably, but I guess it's not worth it.

So, instead of keeping tooling in `packages/build`, it will be more a place to for the tooling configuration (there are still many plugins/configs there).

File: package.json
Comment (ravicious) [inline package.json:1]: Could you add a section to web's README.md explaining where to put deps, outlining a few common use cases?

File: web/packages/shared/package.json [hunk 2]
```diff
@@ -13,5 +13,8 @@
     "@gravitational/design": "1.0.0",
     "ace-builds": "1.35.2",
     "highlight-words-core": "^1.2.2"
+  },
+  "devDependencies": {
+    "jest-fail-on-console": "^3.3.0"
```
Comment (ravicious) [inline web/packages/shared/package.json:18]: This is needed only because `setupTests.tsx` is in shared. I feel like it should be in build instead and it doesn't need the `.tsx` extension.

File: web/packages/build/package.json [hunk 3]
```diff
@@ -60,28 +34,15 @@
...
     "rollup-plugin-visualizer": "^5.12.0",
-    "styled-components": "5.3.11",
     "typescript-eslint": "^7.14.1",
-    "vite": "^5.3.2",
     "vite-plugin-wasm": "^3.3.0",
     "vite-tsconfig-paths": "^4.3.2"
   },
   "devDependencies": {
-    "msw-storybook-addon": "^1.8.0",
     "ts-loader": "^9.5.1"
```
Comment (ravicious) [inline web/packages/build/package.json:46]: `ts-loader` is a Webpack leftover I think, no?

**Review by ravicious — APPROVED** [review]
Merging to v15 might be painful, v16 should be close to master though.

> There are no changes to the `yarn.lock` so I didn't miss any dependencies while moving them :)

Nice. 👍

**Comment by avatus** [comment]
If v15 is too painful, I think going to 16 only is fine and we just move on from there. The benefit of going to 15 really is just we can forget about yarn faster. 

I'm ok with either one

**Review by avatus — APPROVED** [review]

**Comment by ravicious** [comment]
> If v15 is too painful, I think going to 16 only is fine and we just move on from there. The benefit of going to 15 really is just we can forget about yarn faster.

Yeah, I agree. Was there any reason you were thinking of v15 Grzegorz?

File: package.json
Comment (gzdunek) [reply package.json:1]: Good idea, I will do it.

File: web/packages/shared/package.json (hunk 2 above)
Comment (gzdunek) [reply web/packages/shared/package.json:18]: Sure, I will move it to build.

File: web/packages/build/package.json (hunk 3 above)
Comment (gzdunek) [reply web/packages/build/package.json:46]: It's still used by storybook https://github.com/gravitational/teleport/blob/089c32094f4773638b5c2183bbeeeb21dd548d39/web/.storybook/main.js#L100

**Comment by gzdunek** [comment]
>Yeah, I agree. Was there any reason you were thinking of v15 Grzegorz?

I wanted to backport this as far as possible so we would get rid of yarn faster, plus backporting dependency updates would be easier. But well, it turned out that the we don't do it for v15 and v14 anyway (except of electron stuff). 
Initially, I wanted to drop only v14, but v15 is also far from `master`, so I would be probably spend a lot of time backporting it for little benefit.
I'm fine with abandoning this idea, I will switch only master and v16 to pnpm.

File: web/README.md [hunk 4]
```diff
@@ -300,6 +300,18 @@ the root of this repository.
...
+For instance, if you need a new dependency for Web UI, add it to `packages/teleport/package.json`. 
+Similarly, if you're adding a new ESLint plugin, include it in `packages/build/package.json`.
+
+However, there are cases where dependencies should be added to the root `package.json`:
+
+1. Project-wide packages, such as `react`.
+2. Web UI Enterprise dependencies (these shouldn't be added to `e/web/teleport/package.json` 
+to avoid generating a different lockfile when e isn't cloned).
+3. Globally run CLI tools, like `prettier`.
+4. Dependencies imported in multiple workspaces, like `react-day-picker`.
```
Comment (ravicious) [inline web/README.md:313]: I feel like points 1 and 4 are kinda the same, no? React is considered "project-wide" only because Yarn doesn't give you the option to enforce a single package version per all workspaces, which I understand that pnpm does. I believe with pnpm it no longer needs to be considered a "project-wide" package, it's just a dep used by multiple workspaces.

Continuing this train of thought, aren't "dependencies imported in multiple workspaces" pretty much just shortcuts? If both teleport and teleterm need React, why couldn't they both define it in its deps? I believe this is exactly what we'd have to do if they were real packages published to a registry.

The only exception would be e-only deps, for which we use the shortcut of adding it to root package.json because any other way would have been more cumbersome.

I'm not necessarily saying that we should take action on this now (as in, add React to every workspace that uses it etc.). But maybe we could drop the concept of project-wide packages from this list, as I believe it's not a concept that's useful anymore.

File: web/README.md [hunk 5]
```diff
@@ -300,6 +300,18 @@ the root of this repository.
...
+Dependencies should generally be added to the specific workspaces that use them. 
+For instance, if you need a new dependency for Web UI, add it to `packages/teleport/package.json`. 
+Similarly, if you're adding a new ESLint plugin, include it in `packages/build/package.json`.
+
+However, there are cases where dependencies should be added to the root `package.json`:
+
+1. Project-wide packages, such as `react`.
+2. Web UI Enterprise dependencies (these shouldn't be added to `e/web/teleport/package.json` 
+to avoid generating a different lockfile when e isn't cloned).
+3. Globally run CLI tools, like `prettier`.
```
Comment (ravicious) [inline web/README.md:312]: What is "globally run" in this context? 😏 Did you mean packages that provide CLI tools which we'd like to run from the root of the repo?

File: web/README.md (hunk 4 above)
Comment (gzdunek) [reply web/README.md:313]: > Continuing this train of thought, aren't "dependencies imported in multiple workspaces" pretty much just shortcuts? If both teleport and teleterm need React, why couldn't they both define it in its deps? I believe this is exactly what we'd have to do if they were real packages published to a registry.

>The only exception would be e-only deps, for which we use the shortcut of adding it to root package.json because any other way would have been more cumbersome.

I looked again into the deps that we have in the root `package.json` and actually almost all of them are imported in `e`...
So perhaps we should say that dependencies imported in `e` should be kept in the root package.json. If a dependency is not imported in `e`, it should be kept in the workspace.
If a package is needed in multiple workspaces, it should be added in all of them.
For example, `events` should be declared in `teleport`, `teleterm` and `design` packages, instead of in the root.
In other words, I'm saying that we should drop points 1. and 4.

On a side note, the latest version of pnpm introduced [`catalogs` feature](https://pnpm.io/catalogs) that allow you to specify the package version globally. It would help with passing the same version to every workspace. OTOH, it still wouldn't solve the problem with e-only deps.

File: web/README.md (hunk 5 above)
Comment (gzdunek) [reply web/README.md:312]: Yes, I will clarify.

File: web/README.md (hunk 4 above)
Comment (gzdunek) [reply web/README.md:313]: So I'm thinking about something like this:

> Dependencies should generally be added to the specific workspaces that use them. 
> For instance, if you need a new dependency for Web UI, add it to `packages/teleport/package.json`. 
> Similarly, if you're adding a new ESLint plugin, include it in `packages/build/package.json`.
> If a dependency is imported in, for example, `packages/shared` and `packages/teleterm` 
> add it to both of them (that's what we'd have to do if they were real packages 
> published to a registry).
> 
> However, there are cases where dependencies should be added not in a workspace, but 
> in the root `package.json`:
> 1. Dependencies imported in `e` code (we can't declare dependencies in `e/web/teleport/package.json`
>    to avoid generating a different lockfile when `e` isn't cloned). 
>    For example, `react` - it is imported in every package (in `e` too), so it 
>    needs to be kept in the root.
> 2. CLI tools which are run from the root of the repo, like `prettier`.

Comment (ravicious) [reply web/README.md:313]: That sounds good for now. Catalogs seem like a good solution, but I'd worry if tools like dependabot or GitHub Alerts have support for it.

Comment (gzdunek) [reply web/README.md]: >but I'd worry if tools like dependabot or GitHub Alerts have support for it.

Yeah, probably not in the close future.
